DRAFT-  Capture DNS query metrics

amitvc opened the pull request:
The changes are to introduce DnsQueryLifecycleObserver that is responsible for capturing dns query metrics. Please see https://github.com/line/armeria/issues/1887 for details.

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java:
```diff
@@ -0,0 +1,78 @@
+/**
+ * A {@link DnsQueryLifecycleObserver} that helps capture custom dns metrics.
+ */
+public class DefaultDnsQueryLifecycleObserver implements DnsQueryLifecycleObserver {
```
Could remove `public` and add `final`?

amitvc replied:
Yes will do.

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserverFactory.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * A {@link DnsQueryLifecycleObserverFactory} factory that helps create DnsQueryLifecycleObserver.
+ */
+public class DefaultDnsQueryLifecycleObserverFactory implements DnsQueryLifecycleObserverFactory {
```
Ditto.
Could remove `public` and add `final`?

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserverFactory.java:
```diff
@@ -0,0 +1,43 @@
+     * Accepts meterRegistry.
+     * @param meterRegistry {@link MeterRegistry} MeterRegistry to capture metrics.
+     */
+    public DefaultDnsQueryLifecycleObserverFactory(MeterRegistry meterRegistry) {
```
Remove public?

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java:
```diff
@@ -0,0 +1,78 @@
+     * @param meterRegistry {@link MeterRegistry} MeterRegistry to capture metrics.
+     * @param question {@link DnsQuestion} DnsQuestion.
+     */
+    public DefaultDnsQueryLifecycleObserver(MeterRegistry meterRegistry, DnsQuestion question) {
```
Remove `public`?

amitvc replied:
Good point. factory and the observer are in same package. so should be visible.

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    public void queryWritten(InetSocketAddress dnsServerAddress, ChannelFuture future) {
+        meterRegistry.counter("dns.query.count", dnsQuestion.name(), dnsQuestion.type().name()).increment();
```
- How about creating `Counter` in the constructor and reuse it?
- I think we can use `failed` and `succeeded` as a tag. Then we don't need to count for total queries. 

You can refer the follwing code:
https://github.com/line/armeria/blob/9061e241961c806951dd855be65fd871131a1835/core/src/main/java/com/linecorp/armeria/internal/common/metric/RequestMetricSupport.java#L258-L260

ikhoon replied:
On second thought `queryWritten` has meaning it self. How about add a tag with `written`?

amitvc replied:
Sounds good I will make the corresponding changes.

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java:
```diff
@@ -0,0 +1,84 @@
+        this.dnsQuestion = question;
+        success = meterRegistry.counter("dns.query", "result", "success");
+        failure = meterRegistry.counter("dns.query", "result", "failure");
+        queryWritten = meterRegistry.counter("dns.query", question.name(), "written");
```
- We can add more metrics that are enumerated in #1887 such as protocol(UDP, TCP) and DNS error code and so on.
- Could you add test cases to verify this feature's correctness?

trustin replied:
- We could also add a tag for the type of the questions by using `DnsQuestion.type()`
- How about adding a tag for the domain names for all meters? I guess the cardinality will not be that bad.

amitvc replied:
@ikhoon Good suggestions. I will implement them.
@trustin I was worried about the cardinality of adding domain names but it depends on the user of the client api. I think i can go ahead and add it. 
Is there a metrics consumer I can use to see if the metrics are showing up? I have a local unit test that I am using and i can verify in debugger that metrics are being populated. I guess I can extend the UT to add assertions etc.

trustin commented on core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserverFactory.java:
```diff
@@ -0,0 +1,43 @@
+
+    private final MeterRegistry meterRegistry;
+
+    /**
+     * Accepts meterRegistry.
+     * @param meterRegistry {@link MeterRegistry} MeterRegistry to capture metrics.
+     */
```
This is fairly obvious looking from the source code. I think you can remove this block if there's nothing more to add.

amitvc commented:
@trustin @ikhoon I addressed most of the comments. I need to add a good UT to ensure the feature is working correctly.  Not sure right now what the UT should cover. Emitting the metrics might not be same each time the UT runs. Any ideas what would you like to see me cover in UT.

trustin commented on core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    private static String getProtocolType(DnsRecordType type) {
+        if (DnsRecordType.OPT.equals(type) ||
+                DnsRecordType.IXFR.equals(type) ||
+                DnsRecordType.AXFR.equals(type)) {
+            return "tcp";
+        }
```
Does `OPT` really mean `tcp`? IIRC, even a UDP record can have it. If there's no way for observer to determine the protocol, then we can remove it from the metrics.

amitvc replied:
You are correct. I think if the packet size is greater than 512 bytes then it will switch to TCP. Don't know if I can figure this inside of the observer. The observer has access to DnsQuestion but that does not give me enough information.

amitvc commented on core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java:
```diff
@@ -0,0 +1,103 @@
+     */
+    DefaultDnsQueryLifecycleObserver(MeterRegistry meterRegistry, DnsQuestion question, MeterIdPrefix prefix) {
+        this.dnsQuestion = question;
+        success = meterRegistry.counter(prefix.name(), "result", "success", "question", question.name());
+        failure = meterRegistry.counter(prefix.name(), "result", "failure", "question", question.name());
+        dnsErrorCode = meterRegistry.counter(prefix.name(), "dns", "errorcodes", "question", question.name());
+        queryWritten = meterRegistry.counter(prefix.name(), question.name(), "written");
+        queryType = meterRegistry.counter(prefix.name(), question.type().name(), question.name());
+        protocolType = meterRegistry.counter(prefix.name(), "protocol", getProtocolType(question.type()),
+                                            "question", question.name());
```
Thanks @trustin I think I understand what you are asking here. Let me fix this.

amitvc replied:
@trustin If I am passing the MeterIdPrefix upon construction of the WebClient should it be passed in as ClientOption or ClientFactoryOption ?

trustin replied:
How about hard-coding it for now, like `armeria.client.dns.queries` ?

amitvc replied:
@trustin I made the necessary changes. I am trying to unit test the code. I see the counters are being added to the MeterRegistry and I see the code in DefaultDnsQueryLifecycleObserver being called which increments the counters. But when I print out the meter registry the count seems to 0. See below -

{armeria.dns.metrics.test#count{name=google.com.,written=A}=0.0, armeria.dns.metrics.test#count{name=tesla.com.,question=tesla.com.,result=failure}=0.0, armeria.dns.metrics.test#count{name=tesla.com.,question=tesla.com.,result=success}=0.0, armeria.client.connections.lifespan#count{protocol=h2}=0.0, armeria.client.connections.lifespan#total{protocol=h2}=0.0, armeria.client.connections.lifespan#max{protocol=h2}=0.0, armeria.dns.metrics.test#count{name=tesla.com.,type=A}=0.0, armeria.dns.metrics.test#count{name=tesla.com.hsd1.va.comcast.net.,protocol=udp,question=tesla.com.hsd1.va.comcast.net.}=0.0, armeria.client.connections.lifespan#count{protocol=h1}=0.0, armeria.client.connections.lifespan#total{protocol=h1}=0.0, armeria.client.connections.lifespan#max{protocol=h1}=0.0, armeria.dns.metrics.test#count{name=google.com.,type=AAAA}=0.0, armeria.dns.metrics.test#count{name=google.com.,question=google.com.,result=success}=0.0, armeria.dns.metrics.test#count{dns=errorcodes,name=tesla.com.hsd1.va.comcast.net.,question=tesla.com.hsd1.va.comcast.net.}=0.0, armeria.dns.metrics.test#count{dns=errorcodes,name=tesla.com.,question=tesla.com.}=0.0, armeria.dns.metrics.test#count{name=tesla.com.,type=CNAME}=0.0, armeria.dns.metrics.test#count{name=tesla.com.hsd1.va.comcast.net.,type=AAAA}=0.0, armeria.dns.metrics.test#count{name=tesla.com.,type=AAAA}=0.0, armeria.dns.metrics.test#count{name=tesla.com.hsd1.va.comcast.net.,type=CNAME}=0.0, armeria.dns.metrics.test#count{name=tesla.com.hsd1.va.comcast.net.,question=tesla.com.hsd1.va.comcast.net.,result=success}=0.0, armeria.dns.metrics.test#count{name=tesla.com.hsd1.va.comcast.net.,question=tesla.com.hsd1.va.comcast.net.,result=failure}=0.0, armeria.dns.metrics.test#count{name=tesla.com.,written=A}=0.0, armeria.dns.metrics.test#count{name=tesla.com.,written=CNAME}=0.0, armeria.dns.metrics.test#count{name=google.com.,written=AAAA}=0.0, armeria.dns.metrics.test#count{name=google.com.,type=A}=0.0, armeria.dns.metrics.test#count{dns=errorcodes,name=google.com.,question=google.com.}=0.0, armeria.dns.metrics.test#count{name=tesla.com.hsd1.va.comcast.net.,written=AAAA}=0.0, armeria.dns.metrics.test#count{name=google.com.,protocol=udp,question=google.com.}=0.0, armeria.dns.metrics.test#count{name=google.com.,question=google.com.,result=failure}=0.0, armeria.dns.metrics.test#count{name=tesla.com.hsd1.va.comcast.net.,written=CNAME}=0.0, armeria.dns.metrics.test#count{name=tesla.com.,written=AAAA}=0.0, armeria.dns.metrics.test#count{name=tesla.com.,protocol=udp,question=tesla.com.}=0.0}

I am investigating this.

amitvc commented:
@trustin  Do you think I should close this pull request. I think I might have messed up the rebase from master. I am happy to open up a new pull request that is easier to review.
